File: 2075.js
const input = require("fs")
  .readFileSync(
    process.platform === "linux" ? "/dev/stdin" : __dirname + "/2075.txt"
  )
  .toString()
  .trim()
  .split("\n");

const N = Number(input.shift());

const arr = input.map((el) => el.split(" ").map(Number));

class Heap {
  constructor() {
    this.heap = [];
  }
  heapifyUp(index) {
    const parentIdx = Math.floor((index - 1) / 2);
    if (parentIdx >= 0 && this.heap[parentIdx] > this.heap[index]) {
      [this.heap[parentIdx], this.heap[index]] = [
        this.heap[index],
        this.heap[parentIdx],
      ];
      this.heapifyUp(parentIdx);
    }
  }
  heapifyDown(index) {
    const len = this.heap.length - 1;
    const left = index * 2 + 1;
    const right = index * 2 + 2;
    let smallest = index;
    if (left < len && this.heap[left] < this.heap[smallest]) {
      smallest = left;
    }
    if (right < len && this.heap[right] < this.heap[smallest]) {
      smallest = right;
    }

    if (index !== smallest) {
      [this.heap[index], this.heap[smallest]] = [
        this.heap[smallest],
        this.heap[index],
      ];
      this.heapifyDown(smallest);
    }
  }
  insert(value) {
    this.heap.push(value);
    this.heapifyUp(this.heap.length - 1);
  }
  extractMin() {
    if (this.heap.length === 1) return this.heap.pop();

    const answer = this.heap[0];
    this.heap[0] = this.heap.pop();
    this.heapifyDown(0);

    return answer;
  }
}
let answer = 0;
const minHeap = new Heap();
for (let i = 0; i < N; i++) {
  for (let j = 0; j < N; j++) {
    minHeap.insert(arr[i][j]);
    if (minHeap.heap.length > N) {
      answer = minHeap.extractMin();
    }
  }
}
console.log(answer);

File: 2075.py
import heapq

class MaxHeap:
    def __init__(self):
        self.heap = []

    def insert(self, value):
        heapq.heappush(self.heap, -value)  # heapq는 최소 힙을 지원하므로 음수로 저장하여 최대 힙으로 활용

    def extractMax(self):
        if not self.heap:
            raise IndexError("Heap is empty")
        return -heapq.heappop(self.heap)  # 저장된 값을 음수로 사용했으므로 다시 양수로 변환하여 반환

if __name__ == "__main__":
    N = int(input())
    
    arr = []
    for _ in range(N):
        row = list(map(int, input().split()))
        arr.append(row)
    
    maxHeap = MaxHeap()
    for i in range(N):
        for j in range(N):
            maxHeap.insert(arr[i][j])
    
    for i in range(N):
        if i == N - 1:
            print(maxHeap.extractMax())
        else:
            maxHeap.extractMax()
